### PR TITLE
feat(D3 plugin): add allowOverlap dataLabels option

### PR DESCRIPTION
**Pull request by kuzmadom**
closes https://github.com/gravity-ui/chartkit/issues/299

### CODE REVIEW

File: src/plugins/d3/renderer/hooks/useShapes/line/index.tsx [hunk 1]
```diff
@@ -48,86 +49,69 @@ export const LineSeriesShapes = (args: Args) => {
...
+            .attr('y', (d) => d.y)
+            .attr('text-anchor', (d) => d.textAnchor)
+            .style('font-size', (d) => d.style.fontSize)
+            .style('font-weight', (d) => d.style.fontWeight || null)
+            .style('fill', (d) => d.style.fontColor || null);
 
         const hoverEnabled = hoverOptions?.enabled;
         const inactiveEnabled = inactiveOptions?.enabled;
 
+        dispatcher.on('hover-shape.line', null);
```
Comment (korvin89) [inline src/plugins/d3/renderer/hooks/useShapes/line/index.tsx:76]: Why we need this line?

File: src/plugins/d3/renderer/hooks/useShapes/line/index.tsx [hunk 2]
```diff
@@ -48,86 +49,69 @@ export const LineSeriesShapes = (args: Args) => {
...
-            const updates: PreparedLineData[] = [];
-            preparedData.forEach((p) => {
-                const hovered = Boolean(hoverEnabled && p.id === selectedSeriesId);
-                if (p.hovered !== hovered) {
-                    p.hovered = hovered;
-                    updates.push(p);
+            lineSelection.datum((d, index, list) => {
+                const elementSelection = select<BaseType, PreparedLineData>(list[index]);
+
+                const hovered = Boolean(hoverEnabled && d.id === selectedSeriesId);
```
Comment (korvin89) [inline src/plugins/d3/renderer/hooks/useShapes/line/index.tsx:83]: Should we add hovered styles to labels as well?

File: src/plugins/d3/renderer/hooks/useChartOptions/y-axis.ts [hunk 3]
```diff
@@ -49,6 +50,16 @@ const getAxisLabelMaxWidth = (args: {axis: PreparedAxis; series: ChartKitWidgetS
     }).maxWidth;
 };
 
+function getAxisMin(axis?: ChartKitWidgetAxis, series?: ChartKitWidgetSeries[]) {
+    const min = axis?.min;
+
+    if (typeof min === 'undefined' && series?.some((s) => s.type === 'bar-x')) {
```
Comment (korvin89) [inline src/plugins/d3/renderer/hooks/useChartOptions/y-axis.ts:56]: Why we need this condition `series?.some((s) => s.type === 'bar-x')`?

File: src/plugins/d3/examples/line/DataLabels.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
```
Comment (korvin89) [inline src/plugins/d3/examples/line/DataLabels.tsx:1]: We do not use this file yet in storybook?

Comment (kuzmadom) [reply src/plugins/d3/examples/line/DataLabels.tsx:1]: the second example of a line graph in the [showcase](https://preview.gravity-ui.com/chartkit/347/?path=/story/plugins-d3-showcase--d-3-showcase-story)

File: src/plugins/d3/renderer/hooks/useChartOptions/y-axis.ts (hunk 3 above)
Comment (kuzmadom) [reply src/plugins/d3/renderer/hooks/useChartOptions/y-axis.ts:56]: It is for this type of visualization that it is important that the axis starts with 0 - this is the only way to build a rectangle with a minimum value. For lines or points it is not so important.

File: src/plugins/d3/renderer/utils/labels.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+function hasOverlappingByX(rect1: LabelData, rect2: LabelData) {
+    const left1 = getLeftPosition(rect1);
+    const right1 = left1 + rect1.size.width;
+    const left2 = getLeftPosition(rect2);
+    const right2 = left2 + rect2.size.width;
+
+    return Math.max(0, Math.min(right1, right2) - Math.max(left1, left2)) > 0;
+}
+
+function hasOverlappingByY(rect1: LabelData, rect2: LabelData) {
```
[the remark this answers is not in the document]
Comment (kuzmadom) [reply src/plugins/d3/renderer/utils/labels.ts:25]: Filter is a bit wrong, it seems to me.
In this function, we do not filter, but only return the fact about the overlapping.

**Review by korvin89 — APPROVED** [review]

**Pull request merged**